File: tile-client/test/rest/LayerServiceTests.js
/*
 * Copyright (c) 2014 Oculus Info Inc.
 * http://www.oculusinfo.com/
 *
 * Released under the MIT License.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 * of the Software, and to permit persons to whom the Software is furnished to do
 * so, subject to the following conditions:

 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.

 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

"use strict";

var path = require('path'),
    assert = require("assert"),
    srcPath = path.join( __dirname, '../../src/js/rest/' ),  
    LayerService = require( srcPath + 'LayerService' ),
    RestTestUtil = require( './RestTestUtil' ),
    API_VERSION = 'v1.0',
    TEST_LAYER = 'layerId',
    TEST_STATE = 'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855';

describe('LayerService', function() {

    describe('#getLayers()', function() {
        it('should make a GET request to the URL "rest/'+API_VERSION+'/layers"', function() {        
            // mock
            global.$ = RestTestUtil.urlRequestMock( 'get', 'rest/'+API_VERSION+'/layers' );
            LayerService.getLayers();
        });
        it('should provide a function to be called upon error', function() {
            // mock
            global.$ = RestTestUtil.errorFunctionMock( 'get' );               
            LayerService.getLayers();
        });
        it('should accept an optional "success" callback function as a final parameter', function() {
            // mock
            global.$ = RestTestUtil.successFunctionMock( 'get', true );
            // with success
            LayerService.getLayers( function() {
            });

            // mock
            global.$ = RestTestUtil.successFunctionMock( 'get', false );
            // without success
            LayerService.getLayers();
        });
    });

    describe('#getLayer()', function() {
        it('should make a GET request to the URL "rest/'+API_VERSION+'/layers/{layerId}"', function() {
            // mock
            global.$ = RestTestUtil.urlRequestMock( 'get', 'rest/'+API_VERSION+'/layers/'+TEST_LAYER );   
            LayerService.getLayer( TEST_LAYER );
        });
        it('should provide a function to be called upon error', function() {
            // mock
            global.$ = RestTestUtil.errorFunctionMock( 'get' );               
            LayerService.getLayer( TEST_LAYER );
        });
        it('should accept an optional "success" callback function as a final parameter', function() {
            // mock
            global.$ = RestTestUtil.successFunctionMock( 'get', true );
            // with success
            LayerService.getLayer( TEST_LAYER, function() {
            });

            // mock
            global.$ = RestTestUtil.successFunctionMock( 'get', false );
            // without success
            LayerService.getLayer( TEST_LAYER );
        });
    });

    describe('#saveLayerState()', function() {
        it('should make a POST request to the URL "rest/'+API_VERSION+'/layers/{layerId}/states"', function() {
            // mock
            global.$ = RestTestUtil.urlRequestMock( 'post', 'rest/'+API_VERSION+'/layers/'+TEST_LAYER+'/states' );              
            LayerService.saveLayerState( TEST_LAYER, {} );
        });
        it('should provide a function to be called upon error', function() {
            // mock
            global.$ = RestTestUtil.errorFunctionMock( 'post' );                
            LayerService.saveLayerState( TEST_LAYER );
        });
        it('should accept a "parameters" object as a second parameter', function() {
            // mock
            global.$ = {
                post: function( url, param ) {
                        assert( param );
                    return {
                        then: function( success ) {
                            success();
                        }
                    };
                }
            };
            LayerService.saveLayerState( TEST_LAYER, {}, function() {
            });
        });
        it('should accept an optional "success" callback function as a final parameter', function() {
            
            // mock
            global.$ = RestTestUtil.successFunctionMock( 'post', true );
            // with success
            LayerService.saveLayerState( TEST_LAYER, {}, function() {
            });

            // mock
            global.$ = RestTestUtil.successFunctionMock( 'post', false );
            // without success
            LayerService.saveLayerState( TEST_LAYER, {} );
        });
    });

    describe('#getLayerState()', function() {
        it('should make a GET request to the URL "rest/'+API_VERSION+'/layers/{layerId}/states/{stateId}"', function() {
            // mock
            global.$ = RestTestUtil.urlRequestMock( 'get', 'rest/'+API_VERSION+'/layers/'+TEST_LAYER+'/states/'+TEST_STATE );            
            LayerService.getLayerState( TEST_LAYER, TEST_STATE );
        });
        it('should provide a function to be called upon error', function() {
            // mock
            global.$ = RestTestUtil.errorFunctionMock( 'get' );                 
            LayerService.getLayerState( TEST_LAYER, TEST_STATE );
        });
        it('should accept an optional "success" callback function as a final parameter', function() {
            
            // mock
            global.$ = RestTestUtil.successFunctionMock( 'get', true );
            // with success
            LayerService.getLayerState( TEST_LAYER, TEST_STATE, function() {
            });

            // mock
            global.$ = RestTestUtil.successFunctionMock( 'get', false );
            // without success
            LayerService.getLayerState( TEST_LAYER, TEST_STATE );
        });
    });

    describe('#getLayerStates()', function() {
        it('should make a GET request to the URL "rest/'+API_VERSION+'/layers/{layerId}/states"', function() {
            // mock
            global.$ = RestTestUtil.urlRequestMock( 'get', 'rest/'+API_VERSION+'/layers/'+TEST_LAYER+'/states' );            
            LayerService.getLayerStates( TEST_LAYER );
        });
        it('should provide a function to be called upon error', function() {
            // mock
            global.$ = RestTestUtil.errorFunctionMock( 'get' );                 
            LayerService.getLayerStates( TEST_LAYER );
        });
        it('should accept an optional "success" callback function as a final parameter', function() {
            
            // mock
            global.$ = RestTestUtil.successFunctionMock( 'get', true );
            // with success
            LayerService.getLayerStates( TEST_LAYER, function() {
            });

            // mock
            global.$ = RestTestUtil.successFunctionMock( 'get', false );
            // without success
            LayerService.getLayerStates( TEST_LAYER );
        });
    });

});

File: tile-service/src/test/java/com/oculusinfo/tile/rest/layer/LayerServiceTests.java
/*
 * Copyright (c) 2014 Oculus Info Inc. http://www.oculusinfo.com/
 * 
 * Released under the MIT License.
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package com.oculusinfo.tile.rest.layer;

import com.oculusinfo.binning.io.DefaultPyramidIOFactoryProvider;
import com.oculusinfo.binning.io.PyramidIO;
import com.oculusinfo.binning.io.serialization.DefaultTileSerializerFactoryProvider;
import com.oculusinfo.binning.io.serialization.TileSerializer;
import com.oculusinfo.factory.ConfigurationException;
import com.oculusinfo.factory.providers.FactoryProvider;
import com.oculusinfo.tile.init.providers.StandardImageRendererFactoryProvider;
import com.oculusinfo.tile.init.providers.StandardLayerConfigurationProvider;
import com.oculusinfo.tile.init.providers.StandardPyramidIOFactoryProvider;
import com.oculusinfo.tile.init.providers.StandardTilePyramidFactoryProvider;
import com.oculusinfo.tile.init.providers.StandardTileSerializerFactoryProvider;
import com.oculusinfo.tile.init.providers.StandardTileTransformerFactoryProvider;
import com.oculusinfo.tile.rendering.LayerConfiguration;
import com.oculusinfo.tile.rest.QueryParamDecoder;
import com.oculusinfo.tile.rest.config.ConfigException;
import com.oculusinfo.tile.rest.config.ConfigService;
import org.json.JSONObject;
import org.junit.Before;
import org.junit.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.IOException;
import java.net.URISyntaxException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import static org.mockito.Matchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class LayerServiceTests {

    private static final Logger LOGGER = LoggerFactory.getLogger( LayerServiceTests.class );
    private static final String UNIT_TEST_CONFIG_JSON = "unit-test-config.json";

    protected LayerService _layerService;
    private ConfigService _configService;

    @Before
	public void setup () throws Exception {
		try {
			String configFile = "res:///" + UNIT_TEST_CONFIG_JSON;
            Set<FactoryProvider<PyramidIO>> tileIoSet = new HashSet<>();
            tileIoSet.addAll( Arrays.asList( DefaultPyramidIOFactoryProvider.values() ) );
            Set<FactoryProvider<TileSerializer<?>>> serializerSet = new HashSet<>();
            serializerSet.addAll( Arrays.asList( DefaultTileSerializerFactoryProvider.values() ) );
            FactoryProvider<LayerConfiguration> layerConfigurationProvider = new StandardLayerConfigurationProvider(
                new StandardPyramidIOFactoryProvider( tileIoSet ),
                new StandardTilePyramidFactoryProvider(),
                new StandardTileSerializerFactoryProvider(serializerSet),
                new StandardImageRendererFactoryProvider(),
                new StandardTileTransformerFactoryProvider()
            );

            _configService = mock(ConfigService.class);
            withMockConfigService();

            _layerService = new LayerServiceImpl( configFile, layerConfigurationProvider, _configService);
		} catch (Exception e) {
			throw e;
		}
	}

    private void withMockConfigService() throws URISyntaxException, IOException, ConfigException {
        File configFile = new File(this.getClass().getClassLoader().getResource(UNIT_TEST_CONFIG_JSON).toURI());
        String configFileContent = new String(Files.readAllBytes(Paths.get(configFile.getPath())), StandardCharsets.UTF_8);
        when(_configService.replaceProperties(any(File.class))).thenReturn(configFileContent);
    }

    @Test
	public void getLayerJSONsTest() {
		List< JSONObject > jsons = _layerService.getLayerJSONs();
		assert( jsons.size() == 2 );
	}

	@Test
	public void getLayerJsonTest() {
		JSONObject layer0 = _layerService.getLayerJSON( "test-layer0" );
		JSONObject layer1 = _layerService.getLayerJSON( "test-layer1" );
		assert( layer0 != null );
		assert( layer1 != null );
	}

	@Test
	public void getLayerIdsTest() {
		List< String > layerIds = _layerService.getLayerIds();
		assert( layerIds.get(0).equals( "test-layer0" ) );
		assert( layerIds.get(1).equals( "test-layer1" ) );
	}

	@Test
	public void getLayerConfigurationTest() throws ConfigurationException {
		LayerConfiguration layerConfig0 = _layerService.getLayerConfiguration( "test-layer0", null );
		LayerConfiguration layerConfig1 = _layerService.getLayerConfiguration( "test-layer1", null );
		System.out.println( layerConfig0.getPropertyValue( LayerConfiguration.LAYER_ID ) );
		System.out.println( layerConfig1.getPropertyValue( LayerConfiguration.LAYER_ID ) );
		assert( layerConfig0.getPropertyValue( LayerConfiguration.LAYER_ID ).equals( "test-layer0" ) );
		assert( layerConfig1.getPropertyValue( LayerConfiguration.LAYER_ID ).equals( "test-layer1" ) );
	}

	@Test
	public void saveAndGetLayerStateTest() {
		try {
			// create overrides
			JSONObject override = QueryParamDecoder.decode( "renderer.ramp=cool&renderer.coarseness=3" );
			String stateId = _layerService.saveLayerState( "test-layer0", override );
			JSONObject overrideState = _layerService.getLayerState( "test-layer0", stateId );
			assert( overrideState.getJSONObject( "renderer" ).getString( "ramp" ).equals( "cool" ) );
			assert( overrideState.getJSONObject( "renderer" ).getString( "coarseness" ).equals( "3" ) );
		} catch ( Exception e ) {
			LOGGER.error( "Error overriding layer config", e );
		}
	}

	@Test
	public void getLayerStatesTest() {
		try {
			// save state
			JSONObject override = QueryParamDecoder.decode( "renderer.ramp=cool&renderer.coarseness=3&valueTransform.type=linear" );
			String stateId = _layerService.saveLayerState( "test-layer0", override );
			// ensure default is that set in config, NOT propertyValue defaults (unless unconfigured)
			JSONObject states = _layerService.getLayerStates( "test-layer0" );
			// default
			assert( states.getJSONObject( "default" ).getJSONObject( "renderer" ).getString( "ramp" ).equals( "hot" ) );
			assert( states.getJSONObject( "default" ).getJSONObject( "valueTransform" ).getString( "type" ).equals( "log10" ) );
			// saved
			assert( states.getJSONObject( stateId  ).getJSONObject( "renderer" ).getString( "ramp" ).equals( "cool" ) );
			assert( states.getJSONObject( stateId ).getJSONObject( "renderer" ).getString( "coarseness" ).equals( "3" ) );
			assert( states.getJSONObject( stateId ).getJSONObject( "valueTransform" ).getString( "type" ).equals( "linear" ) );
		} catch ( Exception e ) {
			LOGGER.error( "Error overriding layer config", e );
		}
	}

}
